Add make_manifest for openstack

7CHANHO opened the pull request:
Related #31

geofffranks commented on templates/make_manifest:
```diff
@@ -7,6 +7,7 @@ STEMCELL_OS=${STEMCELL_OS:-ubuntu}
 
 infrastructure=$1
 servers=$2
+cf_manifest=$3
```
What values are being pulled from the CF manifest for this? Is it something we can set placeholders for and have people override in their stubs?

7CHANHO replied:
I added 'consul_agent' and 'metron_agent' to 'jobs.public_ha_proxy_z1.templates' because I thought 'consul_agent' and 'metron_agent' jobs are needed. So I needed the properties of the cf manifest. But sadly, I added two jobs and it did not deployed normally.
Sorry for the confusion. I'll try to verify it again.

geofffranks replied:
This haproxy boshrelease is separate from the one included with CF, and designed to be a standalone system if necessary. If you want to use it in place of cf-release's haproxy, you can add the consul_agent, and metron_agent jobs from cf-release, and provide the required properties. Alternatively you could modify the haproxy properties in your CF manifest, and change the jobs for its haproxy to come from the haproxy release, instead of cf.

geofffranks commented on templates/stub.yml:
```diff
@@ -13,8 +15,20 @@ meta:
+  floatings:
+    ha_proxy:
+      public_z1: 172.16.2.6
+
```
I think this should be a list of placeholder IPs, since not everyone will have just one IP, or use that network. Or maybe all the changes in stub.yml should be excluded, since it was built as a default that would get overridden for people as they use it?

geofffranks commented:
Also, can you resolve the merge conflicts that occurred as a result of our repo rename? (Sorry about that :))

geofffranks commented:
@shoudusse since you added the openstack templates for this originally, did you want to take a look at/weigh in on these changes as well?

7CHANHO commented:
Thanks for your review :)
I want to tell you the process I've been through.
Before modifying the source code, I performed the following steps.
Once I have wrote the stub file and ran the make file, I got the following error:
```
$ ./templates/make_manifest openstack 10.10.0.10
RSA 1024 bit CA certificates are loaded due to old openssl compatibility
RSA 1024 bit CA certificates are loaded due to old openssl compatibility
Acting as user 'admin' on 'my-bosh'
my-bosh openstack bosh-openstack-kvm-ubuntu-trusty-go_agent
2016/11/21 08:30:11 error generating manifest: unresolved nodes:
(( merge || m1.small )) in ./templates/infrastructure-openstack.yml compilation.cloud_properties.instance_type ()
(( meta.floatings.ha_proxy.public_z1 )) in dynaml jobs.[0].networks.[1].static_ips ()
(( merge )) in ./templates/infrastructure-openstack.yml meta.floatings.ha_proxy.public_z1 (meta.floatings.ha_proxy.public_z1)
(( merge )) in ./templates/infrastructure-openstack.yml meta.public_security_groups (meta.public_security_groups)
(( meta.public_security_groups )) in dynaml networks.[0].cloud_properties.security_groups ()
(( merge || m1.small )) in ./templates/infrastructure-openstack.yml resource_pools.[0].cloud_properties.instance_type ()
(( merge || m1.small )) in ./templates/infrastructure-openstack.yml resource_pools.[1].cloud_properties.instance_type ()
```

So I added the required properties to the stub file.
```
---
meta:
  ...
  floatings:
    ha_proxy:
      public_z1: 172.16.1.3
      public_z2: 
      private_z1: 172.16.1.4
      private_z2: 
  instnaces:
    ha_proxy:
      public_z1: 1
      public_z2: 0
      private_z1: 1
      private_z2: 0
...
compilation:
  cloud_properties:
    instance_type: m1.small

resource_pools:
  - name: small_z1
    cloud_properties:
      instance_type: m1.small
  - name: small_z2
    cloud_properties:
      instance_type: m1.small
```
The manifest file was finally completed. However, an error occurred during deployment.
There is another error message related to the version of the stemcell.
```
Director task 81843
  Started preparing deployment > Preparing deployment. Failed: Stemcell 'bosh-openstack-kvm-ubuntu-trusty-go_agent/3232' doesn't exist (00:00:01)

Error 50003: Stemcell 'bosh-openstack-kvm-ubuntu-trusty-go_agent/3232' doesn't exist
```

So once again, I added the stemcell version property to the stub file.
```
---
meta:
  ...
  stemcell:
    name: bosh-openstack-kvm-ubuntu-trusty-go_agent
    version: latest
...
```

I thought it was done, but sadly there was an OpenStack API Conflict error.
```
Director task 81845
  Started preparing deployment > Preparing deployment. Done (00:00:00)

  Started preparing package compilation > Finding packages to compile. Done (00:00:00)

  Started creating missing vms
  Started creating missing vms > public_haproxy_z1/0 ()
  Started creating missing vms > private_haproxy_z1/0 (). Failed: OpenStack API Conflict (Multiple possible networks found, use a Network ID to be more specific.). Check task debug log for details. (00:00:02)
   Failed creating missing vms > public_haproxy_z1/0 (): OpenStack API Conflict (Multiple possible networks found, use a Network ID to be more specific.). Check task debug log for details. (00:00:02)
   Failed creating missing vms (00:00:02)

Error 100: OpenStack API Conflict (Multiple possible networks found, use a Network ID to be more specific.). Check task debug log for details.
```
So I added attributes related to the network.
```
networks:
  - name: floating
    type: vip
    cloud_properties:
      security_groups: cf
      net_id: NET-ID
  - name: haproxy1
    type: dynamic
    cloud_properties:
      security_groups: cf
      net_id: NET-ID
  - name: haproxy2
    type: dynamic
    cloud_properties:
      security_groups: cf
      net_id: NET-ID
```

All of these errors are caused by merging each property in 'infrastructure-openstack'.
We can solve the problem by putting the default property value without merge, but I think it is flexible to set it via stub.
What do you think?

geofffranks commented:
I think having it merge and override net ID with what's present in the stub is probably best.

Might make sense to have the stemcell version set to latest in the infrastructure-openstack templates.

Does `(( merge | "m1.small" ))` fix the merge issues, and allow you to not need to override `instance_type` in your stub?

7CHANHO commented:
Thanks for your review :)
As your advice, I do not need to override `instance_type` in the stub anymore after switching to `(( merge | "m1.small" ))`.
I have corrected this, so please review it again.

shinji62 commented:
I think we should go the bosh 2.0 ish.. and stop making this release IaaS dependent.
Maybe except for bosh-lite.

geofffranks commented:
Yeah, I like that approach. I'll put something together

7CHANHO commented:
The bosh 2.0 manifest is now supported! I close this pull request.
